sentinelhub/evalscript.py

```diff
@@ -0,0 +1,117 @@
+"""
+Module defining evalscript generation utilities
+"""
+from typing import Dict, List, Optional
+
+import numpy as np
+
+from .data_collections import DataCollection
+from .data_collections_bands import Band, Unit
+
+DTYPE_TO_SAMPLE_TYPE: Dict[type, str] = {
+    bool: "UINT8",
+    np.uint8: "UINT8",
+    np.uint16: "UINT16",
+    np.float32: "FLOAT32",
+}
+
+EVALSCRIPT_TEMPLATE = """
+    //VERSION=3
+
+    function setup() {{
+        return {{
+            input: [{{
+                bands: [{input_names}],
+                units: [{input_units}]
+            }}],
+            output: [{output_spec}]
+        }}
+    }}
+
+    function updateOutputMetadata(scenes, inputMetadata, outputMetadata) {{
+        outputMetadata.userData = {{
+            "norm_factor":  inputMetadata.normalizationFactor
+        }}
+    }}
+
+    function evaluatePixel(sample) {{
+        return {{ {return_spec} }};
+    }}
+"""
+
+
+def parse_data_collection_bands(data_collection: DataCollection, bands: List[str]) -> List[Band]:
+    """Checks that all requested bands are available and returns the band information for further processing
+
+    :param data_collection: A collection of requested satellite data.
+    :param bands: A list of band or meta band names to use in the evalscript.
+    """
+
+    requested_bands = []
+    band_info_dict = {band_info.name: band_info for band_info in data_collection.bands + data_collection.metabands}
+
+    for band_name in bands:
+        if band_name not in band_info_dict:
+            raise ValueError(
+                f"Data collection {data_collection} does not have specifications for band {band_name}.\n"
+                f"Available bands are:\n{[band.name for band in data_collection.bands]}\nand meta-bands:\n"
+                f"{[band.name for band in data_collection.metabands]}."
+            )
+        requested_bands.append(band_info_dict[band_name])
+
+    return requested_bands
+
+
+def generate_evalscript(
+    data_collection: DataCollection,
+    bands: Optional[List[str]] = None,
+    meta_bands: Optional[List[str]] = None,
+    merged_output: Optional[str] = None,
+    use_dn: bool = True,
+) -> str:
+    """Generate an evalscript based on the provided specifications. This utility supports generating only evalscripts
+    with the mosaicking option set to `SIMPLE`.
```

Comment:
ah, this would also not work with the current SHIT API

Reply:
in SHIT this made sense only in cases where the provided evalscript was a custom one and not the autogenerated one, I believe. Not even sure why the custom evalscript is an option in the first place, isn't SHET supposed to take care of that functionality already?

Reply:
no clue :upside_down_face: its why i'm so keen on replacing it with something better

what precisely would change for other mosaicking options?

Reply:
Ah, wait, I effed up. There is `mosaicking` and `mosaicking_order`:
- the first one controls the flow of data in the evalscript (either it's a single mosaic for a given time interval, or the samples are split per orbit/tile)
  - in generated evalscripts this is always set to SIMPLE, but in custom evalscripts this can be set to anything else, but not sure why this should be supported in SHIT, I would remove custom evalscripts from SHIT, since we have SHET for that.
- the `mosaicking_order` parameter only controls the order in which the data is tiled, (most recent, least recent, least cc..). this is not controlled via the evalscript, but provided in the request payload separately, so from this point of view nothing changes.